lens-data/lens/private/syntax/stx.rkt

```diff
@@ -36,7 +36,19 @@
 ;;                       [Any Any -> Any])
 (define (restore-stx stx dat)
   (if (syntax? stx)
-      (datum->syntax stx dat stx stx)
+      ;; Preserve the distinction between #'(a . (b c)) and #'(a b c)
+      (let loop ([stx1 stx]
+                 [dat1 dat])
+        (cond
+          [(syntax? stx1)
+           (datum->syntax stx1 (loop (syntax-e stx1) dat1) stx1 stx1)]
+          [(and (pair? stx1) (pair? dat1))
```

Comment:
Nit: if it's under 80 columns, I prefer putting the condition and result expression of a `cond` clause on the same line.